x-pack/solutions/observability/plugins/apm/common/service_map/group_resource_nodes.ts

```diff
@@ -0,0 +1,168 @@
+};
+
+function groupConnections({
+  edgesMap,
+  groupableNodeIds,
+}: {
+  edgesMap: Map<string, ConnectionElement>;
+  groupableNodeIds: Set<string>;
+}) {
+  const groupedTargets = Array.from(edgesMap.values()).reduce(
+    (acc, { data: { source, target } }) => {
+      if (groupableNodeIds.has(target)) {
+        const sources = acc.get(target) ?? [];
+        sources.push(source);
+        acc.set(target, sources);
+      }
+
+      return acc;
+    },
+    new Map<string, string[]>()
+  );
+
+  const adjacencyList = Array.from(groupedTargets.entries()).map(([target, sources]) => ({
+    target,
+    sources,
+    groupId: `resourceGroup{${[...sources].sort().join(';')}}`,
+  }));
+
+  const grouped = groupBy(adjacencyList, 'groupId');
+  return Object.entries(grouped)
+    .map(([id, group]) => ({
+      id,
+      sources: group[0].sources,
+      targets: group.map(({ target }) => target),
+    }))
+    .filter(({ targets }) => targets.length >= MINIMUM_GROUP_SIZE);
+}
+
+function getUngroupedNodesAndEdges({
+  nodesMap,
+  edgesMap,
+  groupedConnections,
+}: {
+  nodesMap: Map<string, { data: ConnectionNode }>;
+  edgesMap: Map<string, { data: ConnectionEdge }>;
+  groupedConnections: ReturnType<typeof groupConnections>;
+}) {
+  const ungroupedEdges = new Map(edgesMap);
+  const ungroupedNodes = new Map(nodesMap);
+
+  groupedConnections.forEach(({ sources, targets }) => {
+    targets.forEach((target) => {
+      ungroupedNodes.delete(target);
+      sources.forEach((source) => {
+        ungroupedEdges.delete(getEdgeId({ source, target }));
+      });
+    });
+  });
+
+  return {
+    ungroupedNodes: Array.from(ungroupedNodes.values()),
+    ungroupedEdges: Array.from(ungroupedEdges.values()),
+  };
+}
+
+function groupNodes({
+  nodesMap,
+  groupedConnections,
+}: {
+  nodesMap: Map<string, ConnectionElement>;
+  groupedConnections: ReturnType<typeof groupConnections>;
+}) {
+  return groupedConnections.map(
+    ({ id, targets }): GroupedNode => ({
+      data: {
+        id,
+        [SPAN_TYPE]: 'external',
+        label: i18n.translate('xpack.apm.serviceMap.resourceCountLabel', {
+          defaultMessage: '{count} resources',
+          values: { count: targets.length },
+        }),
+        groupedConnections: compact(
+          targets.map((target) => {
+            const targetElement = nodesMap.get(target);
+            return targetElement
+              ? { label: targetElement.data.label || targetElement.data.id, ...targetElement.data }
+              : undefined;
+          })
+        ),
+      },
+    })
+  );
+}
+
+function groupEdges({
+  groupedConnections,
+}: {
+  groupedConnections: ReturnType<typeof groupConnections>;
+}) {
+  return groupedConnections.flatMap(({ id, sources }) =>
+    sources.map(
+      (source): GroupedEdge => ({
+        data: {
+          id: `${source}~>${id}`,
+          source,
+          target: id,
+        },
+      })
+    )
+  );
+}
+
+export function groupResourceNodes({
```

Comment:
A lot of these methods are crying for some actual iterator methods like `.map`/`.filter`, so you don't have to constantly convert into arrays. Code like this makes me miss Rust iterators so much.